cli/src/main/java/org/jboss/as/cli/impl/CLIModelControllerClient.java

```diff
@@ -297,8 +301,22 @@ public void ensureConnected(long timeoutMillis) throws CommandLineException {
                 }
 
                 if (System.currentTimeMillis() - start > timeoutMillis) {
-                    throw new CommandLineException("Failed to establish connection in " + (System.currentTimeMillis() - start)
-                            + "ms", ioe);
+                    if (timeoutOccured) {
+                        throw new CommandLineException("Failed to establish connection in " + (System.currentTimeMillis() - start)
+                                + "ms", ioe);
+                    } else {
+                        timeoutOccured = true;
+                    }
+                }
+                // Only propagate RedirectException at the end of timeout.
```

Comment:
It's unfortunate that this means it can take a long time to connect, if the CliConfig connectTimeout is high. (IIRC there was a request to make that 30 secs by default to avoid some spurious failures, although maybe that was a different timeout.)

I don't have a good solution though; the approach used by CommandContextImpl to analyze the redirect exception isn't possible here as the requested protocol isn't known.

Bit of a corner case anyway; this http/https redirect in combination with "shutdown --restart=true" + a long timeout config.